### PR TITLE
Refactored to run "go test" from within the test runner

### PR DESCRIPTION
**Pull request by ekingery**
Also simplified source code structure, and removed the unused tests.

Of note - due to moving main.go, git makes it appear that I authored / changed the entire file. This is not the case, which is shown by 2535b838ca278e02c0f9867521a3a0d9356d9b12. I am open to leaving main.go in `cmd/testout/` if we want to keep the history in place. I changed it because `testout` didn't seem apt, and in general I appreciate the simplicity of having a flat repo / code structure (while the codebase remains tiny). I embedded the `Report` and `Test` structs and removed the `exreport` package for the same reason.

### CODE REVIEW

**Comment by iHiD** [comment]
(@tehsphinx Not sure if you've met @ekingery, but he's been very active over the last couple of months across Exercism and has started working on the Go side of things for v3 - just to give some context for you!) 🙁

**Comment by ekingery** [comment]
Of note - running `go test` from within the test runner is in preparation for the upcoming changes to parse out the `test_code` field for V3. Discussion here: https://exercism-team.slack.com/archives/CC4UPN5KP/p1607203773004100?thread_ts=1607203773.004100&cid=CC4UPN5KP

**Review by tehsphinx — APPROVED** [review]
LGTM! Don't know all the new rules, so I leave that to @ErikSchierboom.

@ekingery: Love that you are doing some work on Go! ❤️

File: README.md [hunk 1]
```diff
@@ -1,24 +1,22 @@
...
 
 ## Executing the Test Runner
 
-The test runner takes 3 parameters:
-- the exercise `slug`, e.g. `two-fer`
-- the `path` containing the solution to test
-- the `output path` for the test results
+The test runner takes 2 parameters:
+- `input_dir`: the path containing the solution to test
+- `output_dir`: the output path for the test results
```
Comment (ErikSchierboom) [inline README.md:9]: You're inferring the exercise slug from the directory name? If so, this will be problematic when the test runner executes in our production environment, as the directory names will _not_ match the exercise name. I would suggest to take the slug as the first parameter. The Docker command listed below does have this slug:

```bash
docker run -v ~/Exercism/go/gigasecond:/solution exercism/go-test-runner gigasecond /solution /solutio
```

File: run.sh [hunk 2]
```diff
@@ -1,25 +1,16 @@
 #!/bin/sh
 
 # Arguments:
-# $1: exercise slug (currently ignored)
-# $2: path to solution folder (without trailing slash or preceeding ./)
-# $3: path to output directory (without trailing slash or preceeding ./)
+# $1: `slug` - the exercise slug, e.g. `two-fer` (currently ignored)
```
Comment (ErikSchierboom) [inline run.sh:4]: See comment above: the slug should not be ignored

File: main.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+		if line.Test == "" {
+			continue
+		}
+
+		switch line.Action {
+		case "run":
+			tests[line.Test] = &testResult{
+				Name:     line.Test,
+				TestCode: "test_code field is under construction\ncoming soon!",
```
Comment (ErikSchierboom) [inline main.go:181]: :)

**Review by ErikSchierboom — COMMENTED** [review]
LGTM, but I don't know Go :)

I did notice that no tests are being run in CI. I'd strongly suggest having a test suite and running it as part of the CI, as there can be quite some edge cases. Combine that with some possibly tricky code to get the output/test code, I've found that a test suite is invaluable. What I've done for the C# test runner is define a set of sample solutions that have an `expected_results.json` file that should match the output of the test runner.

File: README.md (hunk 1 above)
Comment (tehsphinx) [reply README.md:9]: That's what I mean with "new rules". From the code it looked to me like `slug` was deprecated. Turns out it wasn't.

Comment (ErikSchierboom) [reply README.md:9]: Correct!

**Comment by ekingery** [comment]
> I did notice that no tests are being run in CI. I'd strongly suggest having a test suite and running it as part of the CI...

@ErikSchierboom I couldn't agree more. I am adding new tests for the new functionality in another branch (see 761772a1567eaeb0bae2d6dc2ee9e0c4d2c174cd). I decided to not add tests in this PR for two reasons. 1) the code is not currently test-friendly and 2) I wanted this PR to be "minimally invasive" and not touch existing functionality. The goal of this PR is simply switching from accepting stdin from the shell script to running `go test` internally. As mentioned above, unfortunately git makes it look like I rewrote the whole main.go file because I moved it and extracted the structs, but in actuality the only changes I made to the existing code was to switch from a bytes array to a bytes buffer in 2535b838ca278e02c0f9867521a3a0d9356d9b12.

File: README.md (hunk 1 above)
Comment (ekingery) [reply README.md:9]: I didn't realize the test runner was not currently correct - I thought the only task was to add the `test_code` field.  I can definitely get the output up to spec. Referencing [the comment I just posted below](https://github.com/exercism/go-test-runner/pull/17#issuecomment-740664545), in order to focus on one fix at a time, I'd like to separate this fix into a separate issue / PR. Is that acceptable,  @ErikSchierboom?

**Comment by ErikSchierboom** [comment]
@ekingery That totally makes sense! Just wanted to mention it :)

File: README.md (hunk 1 above)
Comment (iHiD) [reply README.md:9]: This test runner was one of the first three made when we were really prototying things last year. So there have been a few small changes since then, but I think it's probably still very close to being "correct", bar the `test_code` field. So hopefully not **too** much work!

Comment (ekingery) [reply README.md:9]: For sure - getting it up to spec should be a snap compared to extracting the `test_code`. If we can get this PR merged, I'll get the output up to spec in #19 (which is branched off of this branch), or a future PR.
